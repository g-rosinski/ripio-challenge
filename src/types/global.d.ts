declare global {
    interface Window {
        [name: string]: any
    }
}

export {}