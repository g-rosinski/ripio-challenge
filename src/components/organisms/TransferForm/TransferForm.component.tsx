import React from "react"
import { useForm } from "../../../hooks/useForm/useForm.hook"
import { Button, Form } from "../../atoms"
import { TextField, SelectField } from "../../molecules/Fields"
import { TransferFormProps, TransferFormType } from "./TransferForm.types"

export const TransferForm: React.FC<TransferFormProps> = ({
    initialValues,
    currencies,
    editable = true,
    onSubmit,
    validateForm,
}) => {

    const { 
        values, 
        errors, 
        handleOnChangeField, 
        handleOnSubmitForm 
    } = useForm<TransferFormType>({
        initialValues,
        validateForm,
        onSubmit
    })

    return (
        <Form onSubmit={handleOnSubmitForm}>
            <TextField 
                name="receptor"
                label="Destinatario"
                helper="Ingresar mail valido"
                value={values.receptor}
                error={errors.receptor}
                disabled={!editable}
                onChange={handleOnChangeField}
                />
            
            <SelectField 
                name="currency"
                label="Moneda"
                options={currencies}
                onChange={handleOnChangeField}
                disabled={!editable}
            />
            <TextField 
                name="amount"
                label="Importe"
                helper="Ingresar importe valido"
                value={values.amount}
                error={errors.amount}
                onChange={handleOnChangeField}
                disabled={!editable}
            />

            <Button type="submit" disabled={!editable}>Enviar</Button>
        </Form>
    )
}
