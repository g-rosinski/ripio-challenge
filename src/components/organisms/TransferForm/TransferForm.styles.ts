import styled from 'styled-components'

export const CurrencySection = styled.div`
    display: flex;
    justify-content: center;
    width: 100%;
`