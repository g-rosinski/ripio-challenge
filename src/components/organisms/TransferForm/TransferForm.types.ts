import { UseFormProps } from "../../../hooks/useForm/useForm.hook"

export type TransferFormType = {
    emiter:string,
    receptor:string,
    currency: string, 
    amount: string
}

export type TransferFormProps = UseFormProps<TransferFormType> & {
    currencies: string[]
    editable?: boolean,
}