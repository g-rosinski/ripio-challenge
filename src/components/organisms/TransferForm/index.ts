export * from './TransferForm.component'
export * from './TransferForm.types'
export * from './TransferForm.styles'