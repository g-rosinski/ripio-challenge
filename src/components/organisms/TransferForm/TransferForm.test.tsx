import React from 'react'
import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { TransferForm } from './TransferForm.component'
import { TransferFormType } from './TransferForm.types'

describe('<TransferForm/>', () => {

    const currencies = ['BTC', 'ETH', 'USDT', 'DAI'].map(c => ({value: c}))
    const initialValuesForm: TransferFormType = { emiter: 'usuario@mail.com', receptor: '', currency: currencies[0].value, amount: ''}
    const onSubmitMock = jest.fn()
    const validateForm = jest.fn()
    
    const baseProps = {
        initialValues: initialValuesForm,
        onSubmit: onSubmitMock,
        validateForm: validateForm,
        currencies
    }

    beforeEach(() => {
        jest.clearAllMocks()
        validateForm.mockReturnValue({})
    })

    test('Se reenderice correctamente', () => {
        const {container} = render(<TransferForm {...baseProps}/>)
        expect(screen.getByRole('textbox',{name: /emisor/i})).toHaveValue(initialValuesForm.emiter)
        expect(screen.getByRole('textbox',{name: /emisor/i})).toHaveAttribute('readonly')
        expect(screen.getByRole('textbox',{name: /destinatario/i})).toHaveValue(initialValuesForm.receptor)
        expect(screen.getByRole('combobox',{name: /moneda/i})).toHaveValue(initialValuesForm.currency)
        expect(screen.getByRole('textbox',{name: /importe/i})).toHaveValue(initialValuesForm.amount)
        expect(container).toMatchSnapshot()
    })

    test('Al hacer click verificar si no tiene errores debe hacer submit', () => {
        render(<TransferForm {...baseProps}/>)

        userEvent.click(screen.getByRole('button',{name: /transferir/i}))
        expect(validateForm).toBeCalledWith(initialValuesForm)
        expect(onSubmitMock).toBeCalledWith(initialValuesForm)
    })

    test('Al hacer click verificar si tiene errores no debe hacer submit', () => {
        validateForm.mockReturnValue({receptor: "Destinatario invalido"})
        render(<TransferForm {...baseProps} />)

        userEvent.click(screen.getByRole('button', {name: /transferir/i}))
        expect(validateForm).toBeCalledWith(initialValuesForm)
        expect(screen.getByText("Destinatario invalido")).toBeInTheDocument()
        expect(onSubmitMock).not.toBeCalled()
    })

    test('Debe permitir ingresar valor de token inicial', () => {
        render(<TransferForm {...baseProps} initialValues={{...initialValuesForm, receptor: 'usuario@gmail.com'}} />)

        expect(screen.getByRole('textbox',{name: /destinatario/i})).toHaveValue('usuario@gmail.com')
        userEvent.click(screen.getByRole('button',{name: /transferir/i}))
        expect(validateForm).toBeCalledWith({...initialValuesForm, receptor: 'usuario@gmail.com'})
        expect(onSubmitMock).toBeCalledWith({...initialValuesForm, receptor: 'usuario@gmail.com'})
    })

    test('Si la prop editable esta en false no debe permitir ingresar un valor y cambiar el estilo del form', () => {
        render(<TransferForm {...baseProps} editable={false} />)

        const input = screen.getByRole('textbox',{name: /importe/i})
        const button = screen.getByRole('button',{name: /transferir/i})
        expect(input).toHaveValue('')

        userEvent.type(input, '150')
        expect(input).toHaveValue('')

        userEvent.click(button)
        expect(validateForm).not.toBeCalled()
        expect(onSubmitMock).not.toBeCalled()
    })
})