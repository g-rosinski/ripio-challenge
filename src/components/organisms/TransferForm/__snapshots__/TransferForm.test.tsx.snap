// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TransferForm/> Se reenderice correctamente 1`] = `
<div>
  <form
    class="sc-jSUZER fMUTQF"
  >
    <div
      class="sc-gswNZR kGxItT"
    >
      <label
        class="sc-dkrFOg fZIonL"
        for="receptor"
      >
        Destinatario
      </label>
      <input
        class="sc-gKPRtg crDhfD"
        id="receptor"
        name="receptor"
        type="text"
        value=""
      />
      <p
        class="sc-bcXHqe hkdUZn"
      >
        Ingresar mail valido
      </p>
    </div>
    <div
      class="sc-gswNZR kGxItT"
    >
      <label
        class="sc-dkrFOg fZIonL"
        for="currency"
      >
        Moneda
      </label>
      <select
        class="sc-iBYQkv WkIVr"
        id="currency"
      >
        <option
          value="BTC"
        >
          BTC
        </option>
        <option
          value="ETH"
        >
          ETH
        </option>
        <option
          value="USDT"
        >
          USDT
        </option>
        <option
          value="DAI"
        >
          DAI
        </option>
      </select>
    </div>
    <div
      class="sc-gswNZR kGxItT"
    >
      <label
        class="sc-dkrFOg fZIonL"
        for="amount"
      >
        Importe
      </label>
      <input
        class="sc-gKPRtg crDhfD"
        id="amount"
        name="amount"
        type="text"
        value=""
      />
      <p
        class="sc-bcXHqe hkdUZn"
      >
        Ingresar importe valido
      </p>
    </div>
    <button
      class="sc-hLBbgP dDanPL"
      type="submit"
    >
      Enviar
    </button>
  </form>
</div>
`;
